### PR TITLE
Add long diagnostics for E0046 and E0054

### PR DESCRIPTION
**Pull request by nham**
cc #24407

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -12,6 +12,28 @@
...
 
+E0046: r##"
+When trying to make some type implement a trait `Foo`, you must, at minimum,
+provide implementations for all of `Foo`'s required methods (meaning the
+methods that do not have default implementations), as well as any required
+trait items like associated types or constants.
+"##,
+
+E0054: r##"
+It is not allowed to cast to a bool. If you are trying to cast a numeric type
```
Comment (alexcrichton) [inline src/librustc_typeck/diagnostics.rs]: Perhaps this could clarify "It is not allowed to cast a numeric type to a bool"?

Comment (nham) [inline src/librustc_typeck/diagnostics.rs]: From my (admittedly shallow) reading of the code in librustc_typeck/check/cast.rs, it seemed like any casting to bool is disallowed. For example, you can try to cast a bare function to a bool and still E0054.

```rust
fn main() {
    fn foo() -> u8 { 5 }
    let x = foo as bool; // error: cannot cast as `bool`, compare with zero instead [E0054]
}
```

Comment (alexcrichton) [inline src/librustc_typeck/diagnostics.rs]: Aha! Never mind then!

**Pull request merged**